Reword and add uniformity to Standard Library documentation (#4464)
edgedb/edgedb: docs/stdlib/sequence.rst

            }
        }

    A sequence is bound to the scalar type, not to the property, so
    if multiple properties use the same sequence, they will share the same
    counter. For each distinct counter, a separate scalar type that is
    extending :eql:type:`sequence` should be used.

---------

.. eql:function:: std::sequence_next(seq: schema::ScalarType) -> int64

    Returns the next incrementation of a sequence to its next value.

    Sequence advancement is done atomically; each concurrent session and

>     transaction will receive a distinct sequence value:

Reviewer: ```suggestion
    transaction will receive a distinct sequence value.
```
The code block doesn't illustrate this idea in particular, unless I'm misinterpreting.

                  std::sequence_reset( \
                    seq: schema::ScalarType, val: int64) -> int64

>     Returns the initial state of a sequence or a specified ``val``.

Reviewer 2: This should really be described as a "reset" in that the internal counter of the sequence is reset back to zero. In other words, this has side effects

>     Returns the initial state of a sequence or a specified ``val``.

Reviewer: ```suggestion
    Resets a sequence to initial state or a given value, returning the value.
```

Not in love with this one, but I believe it encapsulates everything and fits within our space constraint.

    The single-parameter form resets the sequence to its initial state, where
    the next :eql:func:`sequence_next` call will return the first value in

>     the sequence. The two-parameters form allows changing the last returned
>     value of the sequence:

Reviewer: ```suggestion
    the sequence. The two-parameter form allows you to set the current value of
    the sequence. The next :eql:func:`sequence_next` call will return the value
    after the one you passed to :eql:func:`sequence_reset`.
```

> .. note::

Reviewer: I had suggested some sweeping changes here that, upon reviewing the documentation more closely in context, didn't make a sense. What I would recommend is that, within the sections for `sequence_next` and `sequence_reset`, we add a paragraph specifically calling out this note and why it's relevant and linking to it.

Suggested callouts:

```rest
    See our notes on specifying the sequence to :eql:func:`sequence_next`.
```

and

```rest
    See our notes on specifying the sequence to :eql:func:`sequence_reset`.
```

In each paragraph, `our notes on specifying the sequence` should link to this point in the documentation.

Reviewer 3: Sure, linking like that sounds OK.

>     Any sequence that can be called upon by either :eql:func:`sequence_next`
>     or :eql:func:`sequence_rest` functions are specified by a
>     ``schema::ScalarType`` object. If the sequence argument is known ahead of
>     time and does not change, the recommended way to pass it is to use the
>     :eql:op:`introspect` operator:

Reviewer: I like what you're trying to convey here — that the note applies to both this and the `sequence_next` function. Instead of doing it this way, though, I'd rather just write a note that applies to one and replicate it for the other. This way, if someone searches only for the `sequence_next` function in our documentation and doesn't read on to any other functions, they'll still have the benefit of this information.

The phrasing "Any sequence that can be called upon" seems a bit convoluted though, so I'll suggest a rewrite that can be repurposed as a note for `sequence_next` swapping out references to this function for the other one.

```suggestion
    The sequence to reset is specified using a ``schema::ScalarType`` object.
    If the sequence argument is known ahead of time and does not change, we
    recommended you pass it by using the :eql:op:`introspect` operator.
```

Rewrite for `sequence_next`:

```rest
.. note::

    The sequence to increment is specified using a ``schema::ScalarType``
    object. If the sequence argument is known ahead of time and does not
    change, we recommended you pass it by using the :eql:op:`introspect`
    operator.
```

>     Any sequence that can be called upon by either :eql:func:`sequence_next`
>     or :eql:func:`sequence_rest` functions are specified by a
>     ``schema::ScalarType`` object. If the sequence argument is known ahead of
>     time and does not change, the recommended way to pass it is to use the
>     :eql:op:`introspect` operator:

Reviewer: ```suggestion
    To specify the sequence to be operated on by either
    :eql:func:`sequence_next` or :eql:func:`sequence_reset`, you must pass a
    ``schema::ScalarType`` object. If the sequence argument is known ahead of
    time and does not change, we recommend passing it by using the
    :eql:op:`introspect` operator:
```

    .. code-block:: edgeql

>         select sequence_next(introspect MySequenceType);
>         # or
>         select sequence_next(introspect typeof MyObj.seq_prop);

Reviewer: Changing all `sequence_next` references to `sequence_reset` to reflect the subject of this section. This content should all be replicated in the `sequence_next` documentation showing that function instead.

```suggestion
        select sequence_reset(introspect MySequenceType);
        # or
        select sequence_reset(introspect typeof MyObj.seq_prop);
```

    This style of execution will ensure that the reference to a sequential
    type from a given expression is tracked properly to guarantee schema
    referential integrity.

>     However, the operation of a sequence type is determined at runtime via. a
>     query argument. It must be queried from the ``schema::ScalarType`` object
>     directly:

Reviewer: ```suggestion
    If the sequence to reset is determined at runtime via a query argument,
    you'll need a different solution. You can query for the sequence from the
    ``schema::ScalarType`` object directly instead:
```

This rewrite no longer reflects the original meaning.

>     However, the operation of a sequence type is determined at runtime via. a
>     query argument. It must be queried from the ``schema::ScalarType`` object
>     directly:

Reviewer: ```suggestion
    It doesn't work in every use case, though. If in your use case, the
    sequence type must be determined at run time via a query argument,
    you will need to query it from the ``schema::ScalarType`` set directly:
```

The rewrite alters the meaning. I have suggested a rework of the original.

    .. code-block:: edgeql

        with
        SeqType := (
            select schema::ScalarType
            filter .name = <str>$seq_type_name
        )
        select

>         sequence_next(SeqType);

Reviewer: ```suggestion
        sequence_reset(SeqType);
```
